src/Raven.Server/Documents/Indexes/IndexStore.cs

```diff
@@ -1004,14 +1005,120 @@ public IEnumerable<Index> GetIndexes()
+        private async Task RunIdleOperationsAsync(long databaseRecordEtag)
         {
-            if (_serverStore.IsLeader() == false)
+            await HandleUnusedAutoIndexes();
+            await DeleteOrMergeSurpassedAutoIndexes(databaseRecordEtag);
+        }
+
+        private async Task DeleteOrMergeSurpassedAutoIndexes(long databaseRecordEtag)
+        {
+            if (_lastSurpassedAutoIndexesDatabaseRecordEtag >= databaseRecordEtag)
                 return;
 
+            _lastSurpassedAutoIndexesDatabaseRecordEtag = databaseRecordEtag;
+
+            var dynamicQueryToIndex = new DynamicQueryToIndexMatcher(this);
+
+            var indexesToRemove = new HashSet<string>();
+            var indexesToExtend = new Dictionary<string, DynamicQueryMapping>();
+
+            foreach (var index in _indexes)
+            {
+                if (index.Type.IsAuto() == false)
+                    continue;
+
+                if (indexesToRemove.Contains(index.Name))
+                    continue;
+
+                var collection = index.Collections.First();
+
+                var query = DynamicQueryMapping.Create(index);
+
+                foreach (var indexToCheck in _indexes.GetForCollection(collection))
+                {
+                    if (index.Type != indexToCheck.Type)
+                        continue;
+
+                    if (index == indexToCheck)
+                        continue;
+
+                    if (indexesToRemove.Contains(indexToCheck.Name))
+                        continue;
+
+                    var definitionToCheck = (AutoIndexDefinitionBase)indexToCheck.Definition;
+
+                    var result = dynamicQueryToIndex.ConsiderUsageOfIndex(query, definitionToCheck);
+                    if (result.MatchType == DynamicQueryMatchType.Complete || result.MatchType == DynamicQueryMatchType.CompleteButIdle)
+                    {
+                        indexesToRemove.Add(index.Name);
+                        indexesToExtend.Remove(index.Name);
+                        break;
+                    }
+
+                    if (result.MatchType == DynamicQueryMatchType.Partial)
+                    {
+                        if (indexesToExtend.TryGetValue(index.Name, out var mapping) == false)
+                            indexesToExtend[index.Name] = mapping = DynamicQueryMapping.Create(index);
+
+                        mapping.ExtendMappingBasedOn(definitionToCheck);
+                    }
+                }
+            }
+
+            if (indexesToRemove.Count == 0 && indexesToExtend.Count == 0)
+                return;
+            
+            foreach (var kvp in indexesToExtend)
+            {
+                var definition = kvp.Value.CreateAutoIndexDefinition();
+
+                if (string.Equals(definition.Name, kvp.Key, StringComparison.Ordinal))
+                    continue;
+
+                try
+                {
+                    await CreateIndex(definition);
```

Comment:
I wonder if it isn't safer to do this one index at a time, instead of all the matches.